pkg/server/plugin/datastore/sql/sql.go

```diff
@@ -1124,149 +1122,761 @@ func createRegistrationEntry(tx *gorm.DB,
+ORDER BY selector_id, dns_name_id
+;`
+	return query, []interface{}{req.EntryId}, nil
+}
+
+func listRegistrationEntries(ctx context.Context, dbType string, db *sql.DB, req *datastore.ListRegistrationEntriesRequest) (*datastore.ListRegistrationEntriesResponse, error) {
+	if req.Pagination != nil && req.Pagination.PageSize == 0 {
+		return nil, status.Error(codes.InvalidArgument, "cannot paginate with pagesize = 0")
+	}
+	if req.BySelectors != nil && len(req.BySelectors.Selectors) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "cannot list by empty selector set")
+	}
+
+	for {
+		resp, err := listRegistrationEntriesOnce(ctx, dbType, db, req)
+		if err != nil {
+			return nil, err
+		}
+
+		if req.BySelectors == nil || len(resp.Entries) == 0 {
+			return resp, nil
+		}
+
+		// Exact/subset matching requires filtering out all registration
+		// entries returned by the query whose selectors are not fully
+		// represented in the request selectors.
+		// TODO: we could probably push this into SQL with some fancy "group
+		// by" count filtering.
+		type selectorKey struct {
+			Type  string
+			Value string
+		}
+		set := make(map[selectorKey]struct{}, len(req.BySelectors.Selectors))
+		for _, s := range req.BySelectors.Selectors {
+			set[selectorKey{Type: s.Type, Value: s.Value}] = struct{}{}
+		}
+		filtered := make([]*common.RegistrationEntry, 0, len(resp.Entries))
+	entryLoop:
+		for _, entry := range resp.Entries {
+			for _, s := range entry.Selectors {
+				if _, ok := set[selectorKey{Type: s.Type, Value: s.Value}]; !ok {
+					continue entryLoop
+				}
 			}
-		case datastore.BySelectors_MATCH_EXACT:
-			selectorsList = append(selectorsList, selectorSet.Raw())
-		default:
-			return nil, fmt.Errorf("unhandled match behavior %q", req.BySelectors.Match)
+			filtered = append(filtered, entry)
+		}
+		resp.Entries = filtered
+
+		if len(resp.Entries) > 0 || resp.Pagination == nil || len(resp.Pagination.Token) == 0 {
+			return resp, nil
 		}
+
+		req.Pagination = resp.Pagination
 	}
+}
 
-	// filter registration entries
-	entryTx := tx
-	if req.ByParentId != nil {
-		entryTx = entryTx.Where("parent_id = ?", req.ByParentId.Value)
+func listRegistrationEntriesOnce(ctx context.Context, dbType string, db *sql.DB, req *datastore.ListRegistrationEntriesRequest) (*datastore.ListRegistrationEntriesResponse, error) {
+	query, args, err := buildListRegistrationEntriesQuery(dbType, req)
+	if err != nil {
+		return nil, sqlError.Wrap(err)
 	}
-	if req.BySpiffeId != nil {
-		entryTx = entryTx.Where("spiffe_id = ?", req.BySpiffeId.Value)
+
+	rows, err := db.QueryContext(ctx, query, args...)
+	if err != nil {
+		return nil, sqlError.Wrap(err)
 	}
+	defer rows.Close()
 
-	if len(selectorsList) == 0 {
-		// no selectors to filter against.
-		var entries []RegisteredEntry
-		entries, p, err = findRegisteredEntries(entryTx, req.Pagination)
-		if err != nil {
-			return nil, sqlError.Wrap(err)
+	var entries []*common.RegistrationEntry
```

Comment:
We'll need a strategy for pre-allocating this in the non-paginated case, this is going to cause a lot of allocations.

Reply:
I feel like I'm in the minority w.r.t. deployment size, and my entire dataset is <100mb right now. I don't think we'll ever be talking about that much memory :)